kotlin/export-tiles/src/main/java/com/example/exporttiles/MainActivity.kt

```diff
@@ -0,0 +1,229 @@
+import com.esri.arcgisruntime.tasks.tilecache.ExportTileCacheTask
+import kotlinx.android.synthetic.main.activity_main.*
+import kotlinx.android.synthetic.main.dialog_layout.*
+import java.io.File
+
+class MainActivity : AppCompatActivity() {
+
+  private val TAG: String = MainActivity::class.java.simpleName
+
+  override fun onCreate(savedInstanceState: Bundle?) {
+    super.onCreate(savedInstanceState)
+    setContentView(R.layout.activity_main)
+
+    // create an ArcGISTiledLayer to use as the basemap
+    val tiledLayer = ArcGISTiledLayer(getString(R.string.world_street_map))
+    val map = ArcGISMap().apply {
+      basemap = Basemap(tiledLayer)
+      minScale = 10000000.0
+    }
+
+    // create a graphic and graphics overlay to show a red box around the tiles to be downloaded
+    val downloadArea = Graphic()
+    downloadArea.symbol = SimpleLineSymbol(SimpleLineSymbol.Style.SOLID, Color.RED, 2f)
+    val graphicsOverlay = GraphicsOverlay()
+    graphicsOverlay.graphics.add(downloadArea)
+
+    mapView.apply {
+      // set the map to the map view
+      this.map = map
+      setViewpoint(Viewpoint(51.5, 0.0, 10000000.0))
+      // add the graphics overlay to the map view
+      graphicsOverlays.add(graphicsOverlay)
+    }
+
+    // update the box whenever the viewpoint changes
+    mapView.addViewpointChangedListener {
+      if (mapView.map.loadStatus == LoadStatus.LOADED) {
+        // upper left corner of the downloaded tile cache area
+        val minScreenPoint: android.graphics.Point = android.graphics.Point(150, 175)
+        // lower right corner of the downloaded tile cache area
+        val maxScreenPoint: android.graphics.Point = android.graphics.Point(
+          mapView.getWidth() - 150,
+          mapView.getHeight() - 250
+        )
+        // convert screen points to map points
+        val minPoint: Point? = mapView?.screenToLocation(minScreenPoint)
+        val maxPoint: Point? = mapView?.screenToLocation(maxScreenPoint)
+        // use the points to define and return an envelope
+        downloadArea.geometry = Envelope(minPoint, maxPoint)
+      }
+    }
+
+    // create up a temporary directory in the app's cache for saving exported tiles
+    val exportTilesDirectory = File(cacheDir, getString(R.string.tile_cache_folder))
+
+    // when the button is clicked, export the tiles to a temporary file
+    exportTilesButton.setOnClickListener {
+      val exportTileCacheTask = ExportTileCacheTask(tiledLayer.uri)
+      val parametersFuture: ListenableFuture<ExportTileCacheParameters> =
+        exportTileCacheTask.createDefaultExportTileCacheParametersAsync(
+          downloadArea.geometry,
+          mapView.mapScale,
+          tiledLayer.maxScale
+        )
+
+      parametersFuture.addDoneListener {
+        val parameters: ExportTileCacheParameters = parametersFuture.get()
+        // export tiles to temporary cache on device
+        val exportTileCacheJob =
```

Comment:
I think this is a good opportunity to use a scoping function to hold all the various listeners:

```
 val parameters: ExportTileCacheParameters = parametersFuture.get()
        // export tiles to temporary cache on device
        val exportTileCacheJob =
          exportTileCacheTask.exportTileCache(parameters, exportTilesDirectory.path + "file.tpk")
            .apply {

              // show progress of the export tile cache job on the progress bar
              addProgressChangedListener {
                dialog.progressBar.progress = progress
              }

              addJobDoneListener {
                dialog.dismiss()
                if (status == Job.Status.SUCCEEDED) {
                  val exportedTileCacheResult: TileCache = result
                  showMapPreview(exportedTileCacheResult)
                } else {
                  ("Job did not succeed: " + error.additionalMessage).also {
                    Toast.makeText(this@MainActivity, it, Toast.LENGTH_LONG).show()
                    Log.e(TAG, error.additionalMessage)
                  }


                }

              }
            }
        // create and show a progress dialog to show export tile progress
        dialog = createProgressDialog(exportTileCacheJob)
        dialog.show()
```

or something like that